### DISCUSSION TITLE
Auto Enroll new resources

### DISCUSSION THREAD

**Discussion by calebcall** [gravitational/teleport#39154] [Q&A] [answered]
Is it possible to have resources auto-enroll?  We primarily run all bare metal servers, so the auto-discovery/enroll mentioned [here](https://goteleport.com/docs/auto-discovery/servers/) won't work.  I was hoping there was a way to include something in our kickstart/post-install process that could join resources.  

We're currently using Teleport Cloud.

**Comment by calebcall** [comment]
Continuing to dig on this, it would seem that maybe we could configure a static token and use that in a script that gets ran on server setup, right?

**Comment by webvictim** [comment, marked as answer]
Where are your resources running? There are lots of better alternatives to using join tokens - like AWS IAM joining, GCP, Azure, Kubernetes etc. You can also use Teleport's auto-discovery features in these environments to automatically register new nodes which come online.

If you're running on-premise then join tokens are probably the easiest way for now. It's not possible to configure static tokens on Teleport Cloud as it requires access to the Teleport config file. This isn't a huge issue, as static tokens are more insecure anyway due to their perpetual validity. We'd recommend using a join token with a shorter length (say, 1 week) and rotating it regularly. You can then pass that into your server bootstrapping process, or find a way for servers to query the latest token from a central source of truth.

`tctl tokens add --type=node --ttl 168h` will give you a 1 week join token for SSH nodes.

It's also possible to generate join tokens via Teleport API if you want to do this a little more dynamically: https://goteleport.com/docs/api/automatically-register-agents/#step-45-write-the-client-application

**Reply by calebcall** [reply]
Thanks, I'll take a look at join tokens. All our servers are bare metal not in a cloud provider. We have data centers around the globe that we colocate in.
